File: Problem2.py
#Time Complexity: O(N)
#Space Complexity: O(N)
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reverseKGroup(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        count = 0
        dummy = ListNode(-1)
        dummy.next = head
        start = dummy
        while head != None:
            count +=1
            if (count%k == 0):
                start = self.reverse(start,head.next)
                head = start.next
            else:
                head = head.next
            
        return dummy.next
                
    def reverse(self,start,end):
        prev  = start
        curr = start.next
        fast = curr.next
        first = curr
        while fast!=end:
            curr.next = prev
            prev = curr
            curr = fast
            fast = fast.next
         
        curr.next = prev
        start.next = curr
        first.next = end
        return first
            

File: Problem2.java

// TC: O(N)
// SC: O(1)
public class Problem2 {
    public ListNode reverseKGroup(ListNode head, int k) {
        ListNode dummy = new ListNode(-1);
        dummy.next = head; 

        ListNode start = dummy;
        ListNode curr = head; 
        int count=1;

        while(curr!=null){
            if(count%k == 0){
                start = reverse(start, curr.next);
                curr = start;
            }
            curr = curr.next;
            count++;
        }

        return dummy.next;
    }

    private ListNode reverse(ListNode start, ListNode end){
        ListNode first = start.next;

        ListNode prev = start;
        ListNode curr = start.next;
        while(curr!=end){
            ListNode temp = curr.next;
            curr.next = prev;
            prev = curr;
            curr = temp;
        }
        first.next = end;
        start.next = prev;
        return first;
    }
}
